src/model/transaction/removeRangeFromContentState.js

```diff
@@ -13,30 +13,266 @@
+};
+
+const getNextValidSibling = (
+  block: ?ContentBlockNode,
+  blockMap: BlockMap,
+  originalBlockMap: BlockMap,
+): ?string => {
+  if (!block) {
+    return null;
+  }
+
+  // note that we need to make sure we refer to the original block since this
+  // function is called within a withMutations
+  let nextValidSiblingKey = originalBlockMap
+    .get(block.getKey())
+    .getNextSiblingKey();
+
+  while (nextValidSiblingKey && !blockMap.get(nextValidSiblingKey)) {
+    nextValidSiblingKey =
+      originalBlockMap.get(nextValidSiblingKey).getNextSiblingKey() || null;
+  }
+
+  return nextValidSiblingKey;
+};
+
+const getPrevValidSibling = (
+  block: ?ContentBlockNode,
+  blockMap: BlockMap,
+  originalBlockMap: BlockMap,
+): ?string => {
+  if (!block) {
+    return null;
+  }
+
+  // note that we need to make sure we refer to the original block since this
+  // function is called within a withMutations
+  let prevValidSiblingKey = originalBlockMap
+    .get(block.getKey())
+    .getPrevSiblingKey();
+
+  while (prevValidSiblingKey && !blockMap.get(prevValidSiblingKey)) {
+    prevValidSiblingKey =
+      originalBlockMap.get(prevValidSiblingKey).getPrevSiblingKey() || null;
+  }
+
+  return prevValidSiblingKey;
+};
+
+const updateBlockMapLinks = (
+  blockMap: BlockMap,
+  startBlock: ContentBlockNode,
+  endBlock: ContentBlockNode,
+  originalBlockMap: BlockMap,
+): BlockMap => {
+  return blockMap.withMutations(blocks => {
+    // update start block if its retained
+    transformBlock(startBlock.getKey(), blocks, block =>
+      block.merge({
+        nextSibling: getNextValidSibling(startBlock, blocks, originalBlockMap),
+        prevSibling: getPrevValidSibling(startBlock, blocks, originalBlockMap),
+      }),
+    );
+
+    // update endblock if its retained
+    transformBlock(endBlock.getKey(), blocks, block =>
+      block.merge({
+        nextSibling: getNextValidSibling(endBlock, blocks, originalBlockMap),
+        prevSibling: getPrevValidSibling(endBlock, blocks, originalBlockMap),
+      }),
+    );
+
+    // update start block parent ancestors
+    getAncestorsKeys(startBlock.getKey(), originalBlockMap).forEach(parentKey =>
+      transformBlock(parentKey, blocks, block =>
+        block.merge({
+          children: block.getChildKeys().filter(key => blocks.get(key)),
+          nextSibling: getNextValidSibling(block, blocks, originalBlockMap),
+          prevSibling: getPrevValidSibling(block, blocks, originalBlockMap),
+        }),
+      ),
+    );
+
+    // update start block next - can only happen if startBlock == endBlock
+    transformBlock(startBlock.getNextSiblingKey(), blocks, block =>
+      block.merge({
+        prevSibling: startBlock.getPrevSiblingKey(),
+      }),
+    );
+
+    // update start block prev
+    transformBlock(startBlock.getPrevSiblingKey(), blocks, block =>
+      block.merge({
+        nextSibling: getNextValidSibling(startBlock, blocks, originalBlockMap),
+      }),
+    );
```

Comment:
It looks like this is removing the 'start block' from the chain of siblings. Like so:
**Before:**
```
A <-> StartBlock <-> B
```
to
```
A <-> B
StartBlock
```

Will keep reading - might be missing something.

Reply:
From this example `A <-> StartBlock <-> B`

startBlock.getPrevSiblingKey() = 'A'

we then update the "A" block nextSibling property to the nearest valid next sibbling (which can be either after the removed range or even sometimes within the range if there was a sibling that got retained for having children)

Reply:
Aaah I see. Thanks, it was helpful to have it written out.